std/arith.asm

```diff
@@ -17,18 +23,174 @@ machine Arith(CLK32_31, operation_id){
     operation ec_double<12> x1c[0], x1c[1], x1c[2], x1c[3], x1c[4], x1c[5], x1c[6], x1c[7], y1c[0], y1c[1], y1c[2], y1c[3], y1c[4], y1c[5], y1c[6], y1c[7] -> x3c[0], x3c[1], x3c[2], x3c[3], x3c[4], x3c[5], x3c[6], x3c[7], y3c[0], y3c[1], y3c[2], y3c[3], y3c[4], y3c[5], y3c[6], y3c[7];
     
     let BYTE: col = |i| i & 0xff;
     let BYTE2: col = |i| i & 0xffff;
 
-    pol commit x1[16], y1[16], x2[16], y2[16], x3[16], y3[16], s[16], q0[16], q1[16], q2[16];
+    let secp_modulus = 0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2f;
+
+    let inverse: int -> int = |x| ff::inverse(x, secp_modulus);
+    let add = |x, y| ff::add(x, y, secp_modulus);
+    let sub = |x, y| ff::sub(x, y, secp_modulus);
+    let mul = |x, y| ff::mul(x, y, secp_modulus);
+    let div = |x, y| ff::div(x, y, secp_modulus);
+
+    pol commit x1[16], y1[16], x2[16], y2[16], x3[16], y3[16];
+
+    // Selects the ith limb of x (little endian)
+    // Note that the most significant limb can be up to 32 bits; all others are 16 bits.
+    let select_limb = |x, i| fe((x >> (i * 16)) & if i < 15 { 0xffff } else { 0xffffffff });
+
+    let s_for_eq1 = |x1, y1, x2, y2| div(sub(y2, y1), sub(x2, x1));
+    let s_for_eq2 = |x1, y1| div(mul(3, mul(x1, x1)), mul(2, y1));
+
+    // Adding secp_modulus to make sure that that all numbers are positive when % is applied to it.
+    let compute_x3_int = |x1, x2, s| (s * s - x1 - x2 + 2 * secp_modulus) % secp_modulus;
+    let compute_y3_int = |x1, y1, x3, s| (s * ((x1 - x3) + secp_modulus) - y1 + secp_modulus) % secp_modulus;
+
+    // Compute quotients for the various equation.
+    // Note that we add 2**258 to it, to move it from the (-2**258, 2**258) to the (0, 2**259) range, so it can
+    // be represented as an unsigned 272-bit integer.
+    // See the comment for `product_with_p` below.
+    let compute_q0_for_eq1 = |x1, y1, x2, y2, s| (-(s * x2 - s * x1 - y2 + y1) / secp_modulus + (1 << 258));
+    let compute_q0_for_eq2 = |x1, y1, s| (-(2 * s * y1 - 3 * x1 * x1) / secp_modulus + (1 << 258));
+    let compute_q1 = |x1, x2, x3, s| (-(s * s - x1 - x2 - x3) / secp_modulus + (1 << 258));
+    let compute_q2 = |x1, y1, x3, y3, s| (-(s * x1 - s * x3 - y1 - y3) / secp_modulus + (1 << 258));
+ 
+    let limbs_to_int: expr[] -> int = |limbs| array::sum(array::map_enumerated(limbs, |i, limb| int(eval(limb)) << (i * 16)));
+
+    let x1_int: -> int = || limbs_to_int(x1);
+    let y1_int: -> int = || limbs_to_int(y1);
+    let x2_int: -> int = || limbs_to_int(x2);
+    let y2_int: -> int = || limbs_to_int(y2);
+    let x3_int: -> int = || limbs_to_int(x3);
+    let s_int: -> int = || limbs_to_int(s);
+
+    let eq1_active = || eval(selEq[1]) == 1;
+    let get_operation = || match eval(operation_id) {
+        1 => "affine_256",
+        10 => "ec_add",
+        12 => "ec_double",
+        _ => panic("Unknown operation")
+    };
+    let is_ec_operation: -> int = || match get_operation() {
```

Comment:
This should return a bool, but I couldn't make it work:
- Returning `true` and `false` didn't work (can't find the symbol). Are these literals not available?
- Returning `get_operation() != "affine_256"` didn't work (string doesn't implement equality).

Reply:
You can add the boolean constants as `let true = (1 == 1)` or something.

Reply:
Ah probably `1 > 0`, otherwise the type is not uniquely defined.

Reply:
Hmm, I tried:
```
    let true = 1 >= 1;
    let false = 1 > 1;
```
But this gives me:
```
Error during type inference:
Error specializing generic references in main_arith.true:
Unable to derive concrete type for literal 1.
Error specializing generic references in main_arith.true:
Unable to derive concrete type for literal 1.
Error specializing generic references in main_arith.false:
Unable to derive concrete type for literal 1.
Error specializing generic references in main_arith.false:
Unable to derive concrete type for literal 1.
```